feat(bitbuckerserver): Add support for Bitbucket Server

pierredup opened the pull request:
Add support for Bitbucket server to link commits

Fixes #8372, getsentry/sentry-plugins#197, getsentry/sentry-plugins#383


### TODO

- [x] Unit tests

scefali commented:
@pierredup I am testing you PR locally. I see you have a link to the [required steps ](https://docs.sentry.io/workflow/integrations/global-integrations/bitbucket-server/)to the docs but that link doesn't exist. Could you make a PR in our [docs repo ](https://github.com/getsentry/sentry-docs)with the required steps? This will also help me test and validate your PR locally.

pierredup commented:
@scefali Yes I'm working on the docs, will create the PR as soons as I'm done. For now you can follow the steps for Jira Server, as the steps are very similar with slight adjustements to the settings (E.G replacing all instances of `jira_server` with `bitbucket_server`)

scefali commented on src/sentry/integrations/bitbucket_server/client.py:
```diff
@@ -0,0 +1,139 @@
+
+    def get_repo(self, project, repo):
+        return self.get(
+            "/rest/api/1.0/projects/%s/repos/%s" % (project, repo), auth=self.get_auth()
```
@pierredup Can you create a class that has the path templates as we do with BitBucket Cloud with the `BitbucketAPIPath` class? Would like to be consistent when possible.

scefali commented on src/sentry/integrations/bitbucket_server/repository.py:
```diff
@@ -0,0 +1,113 @@
+                        )
+                    ),
+                    "active": True,
+                    "events": ["repo:refs_changed", "pr:merged"],
```
@pierredup I see for the BitBucket Cloud integration we have hooks for the following events:
```
["repo:push", "pullrequest:fulfilled"]
```
Are they the same events in BitBucket server with different names?

pierredup replied:
Yes it's the same events, but BitBucket server just has different names

scefali commented on src/sentry/integrations/bitbucket_server/repository.py:
```diff
@@ -0,0 +1,113 @@
+            installation.raise_error(e)
+        else:
+            config["external_id"] = six.text_type(repo["id"])
+            config["name"] = repo["project"]["name"] + "/" + repo["name"]
```
@pierredup I see that we are often splitting the name downstream so we can split the project and repo name. This makes it a bit confusing because it might be super clear that `parts[0]` is the project and `parts[1]` is the name in other places in the code. Any reason we can't store the `project` so the name isn't `{project}/{name}`? We could call the latter field `long_name` if we still need it in other places.

scefali commented on src/sentry/integrations/bitbucket_server/repository.py:
```diff
@@ -0,0 +1,113 @@
+        client = installation.get_client()
+        # use config name because that is kept in sync via webhooks
+        parts = repo.config["name"].split("/")
+        try:
```
@pierredup I see in BitBucket Cloud we have a check if `start_sha` is `None`. Wh don't we need the check here?

pierredup replied:
> I see in BitBucket Cloud we have a check if start_sha is None. Why don't we need the check here?

`start_sha` won't ever be `None`, as the hook just sends the value as `0000000000000000000000000000000000000000` when it's empty.

> I noticed that when my first commit (but not subsequent commits) produced an error on the webhook:

This is definitely related, I'll have a look at this and see how we can handle an empty commit

scefali commented on src/sentry/static/sentry/app/plugins/components/pluginIcon.tsx:
```diff
@@ -5,6 +5,7 @@ import placeholder from 'app/../images/integrations/integration-default.png';
 import sentry from 'app/../images/integrations/integration-sentry.png';
 import asana from 'app/../images/integrations/asana-logo.png';
 import bitbucket from 'app/../images/integrations/bitbucket-logo.png';
+import bitbucketserver from 'app/../images/integrations/bitbucket-server-logo.png';
```
@pierredup Can you resize the image to be 72x72? I realize that we aren't following this standard right now but going forward, that is going to be the requirement.

scefali commented:
@pierredup I pulled down and tested the PR and it looks great! Everything is working as expected aside from the one error I mentioned in another comment. As a member of the Ecosystem team at Sentry, I really appreciate the effort you put into making this happen and I know members of the open-source community are thrilled as well. We should be able to get this merged after addressing my comments and getting the tests added. Well done!

scefali commented on src/sentry/integrations/bitbucket_server/repository.py:
```diff
@@ -0,0 +1,124 @@
+        # use config name because that is kept in sync via webhooks
+
+        # Bitbucket servers send an empty commit as a string of zeros
+        if "0000000000000000000000000000000000000000" == start_sha:
```
@pierredup Could we do something like: 
```
"0"*40
```
I think it's a bit cleaner than having to write out 40 `0`s.

scefali commented on src/sentry/integrations/bitbucket_server/webhook.py:
```diff
@@ -0,0 +1,158 @@
+        # while we're here, make sure repo data is up to date
+        self.update_repo_data(repo, event)
+
+        parts = repo.name.split("/")
```
@pierredup Related to my earlier comment about clarity, could we do something like this instead:
```
[project_name, repo_name] = repo.name.split("/")
```
That makes the subsequent invocation of `get_commits` a bit easier to understand.

pierredup commented:
@scefali Thanks for the review, this is my first time working in Python, so it's nice to learn all these neat little 'tricks' :)
I'm working on the tests and the docs, will push them soon.

scefali commented:
@pierredup Any updates on adding tests?

pierredup commented:
@scefali I'm having some issues running the tests locally, just need to sort out some environment issues and will push all the tests then

scefali commented:
@pierredup Understood. If you have any questions, please don't hesitate to reach out. Some recent changes I made have messed with people's environments so I might be able to help. Here's one command that might help:
```
SENTRY_LIGHT_BUILD=1 pip install -e . --no-use-pep517
```

l0nax commented:
@scefali @pierredup Any Updates about this Pull-Request?

ErnestStaug commented:
Anyone still on this?

LouisMilesRH commented:
Any updates on this? We are still super keen to have On Prem bitbucket integration!

scefali commented:
@pierredup @LouisMilesRH Sorry for the delay! We will get back to reviewing this shortly.

scefali commented on tests/sentry/integrations/bitbucket_server/testutils.py:
```diff
@@ -0,0 +1,109 @@
+from __future__ import absolute_import
+from collections import OrderedDict
+
+EXAMPLE_PRIVATE_KEY = """-----BEGIN RSA PRIVATE KEY-----
```
@pierredup Where is this private key from? This looks like a real private key which makes me nervous, even if it's no longer valid. Probably better to make it look faker so no one thinks it's a valid private key :)

pierredup replied:
@scefali I just copied it from [tests/sentry/integrations/jira_server/testutils.py](https://github.com/getsentry/sentry/blob/master/tests/sentry/integrations/jira_server/testutils.py#L3)

scefali replied:
@pierredup
That’s fine then :)

scefali commented:
@pierredup Sorry for taking so long to get back to this after you added unit tests! BTW, everyone at Sentry really appreciates that you took the time to make this PR and we will definitely be sending you something to show you our gratitude.

I just pulled down and tested this from scratch again. When I tested this integration originally in October, everything was working fine except one edge case which you fixed. However, when I retest this, I am consistently getting an error when I try to add a repository. Here's the view:

![Screen Shot 2020-02-14 at 10 18 52 AM](https://user-images.githubusercontent.com/8533851/74556672-7094e300-4f13-11ea-8d9c-32aaa1de767d.png)

And the error:

```
18:10:31 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/repository.py", line 31, in get_repository_data
18:10:31 server  |     repo = client.get_repo(config["project"], config["repo"])
18:10:31 server  | KeyError: 'project'
```

When I print out `config`, I see this:
```
{u'identifier': u'TEST/race', u'installation': u'22', u'provider': u'integrations:bitbucket_server'}
```
This makes sense because you can see that's what the request is providing here: https://github.com/pierredup/sentry/blob/bitbucket-server/src/sentry/static/sentry/app/actionCreators/integrations.jsx#L146

As far as I can tell,`config` is just the request data that's being sent by the front end. I tried debugging to understand where `project` would be set in the config but I am really stumped. Now I am very confused about how I got this working previously but now I cannot! Any idea what's going on?

Side note, I would definitely rebase from master as files have changed since you created your branch. That file I pointed to is now a Typescript file. I don't think anything has changed to impact the functionality but you never know.

pierredup commented:
@scefali I tested again and also got the same error as you. This is strange, as it was also working previously when I tested it, so not sure what happened now. I have fixed the repository to use the identifier, as there is no project key set in the config. I have also rebased from master and tested again and adding a repository now works again as expected.

scefali commented on src/sentry/integrations/bitbucket_server/webhook.py:
```diff
@@ -0,0 +1,158 @@
+
+        for change in event["changes"]:
+            for commit in client.get_commits(
+                project_name, repo_name, change.get("fromHash"), change.get("toHash")
```
@pierredup So I am getting an error when I try to push to a repo for the first time:
```
19:55:02 server  | "POST /api/0/organizations/sentry/repos/ HTTP/1.1" 201 2431
19:55:21 server  | Traceback (most recent call last):
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
19:55:21 server  |     response = get_response(request)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
19:55:21 server  |     response = self._get_response(request)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
19:55:21 server  |     response = self.process_exception_by_middleware(e, request)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
19:55:21 server  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
19:55:21 server  |     return self.dispatch(request, *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper
19:55:21 server  |     return bound_func(*args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
19:55:21 server  |     return view_func(*args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/utils/decorators.py", line 63, in bound_func
19:55:21 server  |     return func.__get__(self, type(self))(*args2, **kwargs2)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/webhook.py", line 116, in dispatch
19:55:21 server  |     return super(BitbucketServerWebhookEndpoint, self).dispatch(request, *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/views/generic/base.py", line 88, in dispatch
19:55:21 server  |     return handler(request, *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/webhook.py", line 157, in post
19:55:21 server  |     handler()(organization, integration_id, event)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/webhook.py", line 68, in __call__
19:55:21 server  |     project_name, repo_name, change.get("fromHash"), change.get("toHash")
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/bitbucket_server/client.py", line 144, in get_commits
19:55:21 server  |     params={"since": fromHash, "until": toHash, "merges": "exclude"},
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/client.py", line 230, in get
19:55:21 server  |     return self.request("GET", *args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/client.py", line 224, in request
19:55:21 server  |     return self._request(*args, **kwargs)
19:55:21 server  |   File "/Users/scefali/Work/sentry/src/sentry/integrations/client.py", line 214, in _request
19:55:21 server  |     raise ApiError.from_response(resp)
19:55:21 server  | ApiError: {"errors":[{"context":null,"message":"Commit '0000000000000000000000000000000000000000' does not exist in repository 'new repo 3'.","exceptionName":"com.atlassian.bitbucket.commit.NoSuchCommitException"}]}
```
The fix is pretty simple though:
```
            from_hash = None if change.get("fromHash") == "0" * 40 else change.get("fromHash")
            for commit in client.get_commits(
                project_name, repo_name, from_hash, change.get("toHash")
            )["values"]:
```

If you can get this fix and it works, I am ready to merge this PR!

pierredup replied:
Thank you @scefali, I have pushed the change. I'll see if I can do a full test and make sure everything still works

scefali commented:
@pierredup
Thanks for taking care of that bug. We are so close to getting this merged! Unfortunately, the build is failing because of a unit test:

BitbucketServerRepositoryProviderTest.test_build_repository_config ______
849Traceback (most recent call last):
850  File "<string>", line 3, in wrapper
851  File "/home/travis/build/getsentry/sentry/tests/sentry/integrations/bitbucket_server/test_repository.py", line 111, in test_build_repository_config
852    data = self.provider.get_repository_data(organization, data)
853  File "/home/travis/build/getsentry/sentry/src/sentry/integrations/bitbucket_server/repository.py", line 33, in get_repository_data
854    installation.raise_error(e)
855  File "/home/travis/build/getsentry/sentry/src/sentry/integrations/base.py", line 338, in raise_error
856    IntegrationError, IntegrationError(self.message_from_error(exc)), sys.exc_info()[2]
857  File "/home/travis/build/getsentry/sentry/src/sentry/integrations/bitbucket_server/repository.py", line 30, in get_repository_data
858    project, repo = config["identifier"].split("/", 1)

scefali commented:
@pierredup So I just pulled down to test the change you made and I am getting another error:

```
Traceback (most recent call last):
  File "/Users/scefali/Work/sentry/.venv/bin/sentry", line 11, in <module>
    load_entry_point('sentry', 'console_scripts', 'sentry')()
  File "/Users/scefali/Work/sentry/src/sentry/runner/__init__.py", line 164, in main
    cli(prog_name=get_prog(), obj={}, max_content_width=100)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/scefali/Work/sentry/src/sentry/runner/decorators.py", line 68, in inner
    return ctx.invoke(f, *args, **kwargs)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/scefali/Work/sentry/src/sentry/runner/decorators.py", line 29, in inner
    configure()
  File "/Users/scefali/Work/sentry/src/sentry/runner/__init__.py", line 127, in configure
    configure(ctx, py, yaml, skip_service_validation)
  File "/Users/scefali/Work/sentry/src/sentry/runner/settings.py", line 158, in configure
    skip_service_validation=skip_service_validation,
  File "/Users/scefali/Work/sentry/src/sentry/runner/initializer.py", line 311, in initialize_app
    django.setup()
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate
    app_config.import_models()
  File "/Users/scefali/Work/sentry/.venv/lib/python2.7/site-packages/django/apps/config.py", line 202, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/local/Cellar/python@2/2.7.16_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/scefali/Work/sentry/src/sentry/plugins/sentry_interface_types/models.py", line 7, in <module>
    from sentry.plugins.bases.tag import TagPlugin
  File "/Users/scefali/Work/sentry/src/sentry/plugins/bases/__init__.py", line 4, in <module>
    from .issue2 import IssueTrackingPlugin2  # NOQA
  File "/Users/scefali/Work/sentry/src/sentry/plugins/bases/issue2.py", line 20, in <module>
    from sentry.plugins.endpoints import PluginGroupEndpoint
  File "/Users/scefali/Work/sentry/src/sentry/plugins/endpoints.py", line 7, in <module>
    from sentry.api.bases.group import GroupEndpoint
  File "/Users/scefali/Work/sentry/src/sentry/api/bases/__init__.py", line 3, in <module>
    from .group import *  # NOQA
  File "/Users/scefali/Work/sentry/src/sentry/api/bases/group.py", line 5, in <module>
    from sentry.api.base import Endpoint
  File "/Users/scefali/Work/sentry/src/sentry/api/base.py", line 29, in <module>
    from sentry.web.api import allow_cors_options
  File "/Users/scefali/Work/sentry/src/sentry/web/api.py", line 30, in <module>
    from sentry_relay import ProcessingActionInvalidTransaction, scrub_event
ImportError: cannot import name ProcessingActionInvalidTransaction
```
But the error goes away when I rebase from master so I would recommend doing that again.

scefali commented:
@pierredup Any chance you could fix the broken test soon? If not, I would be happy to fix things up and create a new branch off yours and make a PR that can be merged.

ErnestStaug commented:
I will probably need a guide because I cannot find on bitbucket server where is `consumer key`

LouisMilesRH commented:
Agreed - this link is broken which is annoying as without it slightly going in blind

https://docs.sentry.io/workflow/integrations/global-integrations/bitbucket-server/

I can see that @scefali mentioned this link right at the top of the thread - is there an updated link we can go to (or even one hosted else where just to get us testing)? @pierredup?

pierredup commented:
I will work on the documentation and open a PR soon, for now you can just follow the Jira Server steps and just replace all instances of Jira with Bitbucket

scefali commented:
@pierredup I would be happy to make the documentation PR, although I won’t be able to get to it until next week.

LouisMilesRH commented:
Would it be possible to allow the BitBucket URL to not contain any '.' characters. In our case the URL is setup simply to be https://bitbucket but it appears that this is failing validation 

![image](https://user-images.githubusercontent.com/59727334/75559333-2d6e6000-5a3b-11ea-8fd9-ff5265376321.png)

scefali commented:
@LouisMilesRH It’s something we could consider. Could you make a Github issue for this request and link it here?

LouisMilesRH commented:
Sure thing! Got it working for now (looking great!) but will add this as a request later :)

scefali commented:
@pierredup BTW, I just started on the docs so no need to worry about them. Hopefully, I'll get a PR up today.

scefali commented:
Doc PR: https://github.com/getsentry/sentry-docs/pull/1519

scuben commented:
using `sentry-cli` with `set-commits --auto` i get:
```
DEBUG   2020-02-29 16:01:01.231677 +01:00   unknown repository integrations:bitbucket_server:57814 (https://xxx.tld/projects/xxxx/repos/xxx/browse) skipped
```

Does this happen to be connected to this PR?

aldanor commented:
Same issue as @LouisMilesRH here -- on-premise BitBucket server URLs would often not contain any dots (e.g. `https://bitbucket/`). @LouisMilesRH how did you solve it?

scuben commented:
It seems there is a problem with this new integration: https://github.com/getsentry/sentry/issues/17398

LouisMilesRH commented:
> 
> 
> Same issue as @LouisMilesRH here -- on-premise BitBucket server URLs would often not contain any dots (e.g. `https://bitbucket/`). @LouisMilesRH how did you solve it?

We ended up using fully qualified server name including domain (so https://bitbucket.domain) which in our case resolved to the same place. Worth noting this is the same restriction as other integrations (we had to do the same thing when connecting to Jira on prem)

pierredup commented:
> using `sentry-cli` with `set-commits --auto` i get:
> 
> ```
> DEBUG   2020-02-29 16:01:01.231677 +01:00   unknown repository integrations:bitbucket_server:57814 (https://xxx.tld/projects/xxxx/repos/xxx/browse) skipped
> ```

I think this is because Bitbucket Server is not supported in the sentry-cli yet ([getsentry/sentry-cli:src/utils/vcs.rs#L201](https://github.com/getsentry/sentry-cli/blob/f691550c3f1b76e2ee7db62bb42f570a129730f7/src/utils/vcs.rs#L201))
